go/p2p/discovery/bootstrap/client.go

```diff
@@ -0,0 +1,162 @@
+package bootstrap
+
+import (
+	"context"
+	"fmt"
+	"math/rand"
+	"sync"
+	"time"
+
+	"github.com/libp2p/go-libp2p/core/discovery"
+	"github.com/libp2p/go-libp2p/core/host"
+	"github.com/libp2p/go-libp2p/core/peer"
+	"github.com/libp2p/go-libp2p/core/peerstore"
+
+	"github.com/oasisprotocol/oasis-core/go/common/logging"
+	"github.com/oasisprotocol/oasis-core/go/p2p/rpc"
+)
+
+// ClientOptions are client's peer discovery options.
+type ClientOptions struct {
+	retentionPeriod time.Duration
+}
+
+// ClientOption is a client peer discovery option setter.
+type ClientOption func(opts *ClientOptions)
+
+// WithRetentionPeriod configures peer retention period.
+func WithRetentionPeriod(retentionPeriod time.Duration) ClientOption {
+	return func(opts *ClientOptions) {
+		opts.retentionPeriod = retentionPeriod
+	}
+}
+
+// client is an implementation of a peer discovery that fetches peers from the seed node and
+// advertises services.
+type client struct {
+	logger *logging.Logger
+
+	host host.Host
+	rc   rpc.Client
+
+	seed            peer.AddrInfo
+	retentionPeriod time.Duration
+
+	mu         sync.Mutex
+	peers      map[string][]peer.AddrInfo
+	timestamps map[string]time.Time
+}
+
+// NewClient creates a new bootstrap protocol client.
+func NewClient(h host.Host, seed peer.AddrInfo, opts ...ClientOption) discovery.Discovery {
+	l := logging.GetLogger("p2p/discovery/bootstrap").With(
+		"seed_id", seed.ID,
+		"seed_addrs", seed.Addrs,
+	)
+
+	do := ClientOptions{}
+	for _, opt := range opts {
+		opt(&do)
+	}
+
+	// Add the seed to the peer store so that libp2p knows how to contact it.
+	h.Peerstore().AddAddrs(seed.ID, seed.Addrs, peerstore.PermanentAddrTTL)
+
+	rc := rpc.NewClient(h, ProtocolID())
+
+	return &client{
+		logger:          l,
+		host:            h,
+		rc:              rc,
+		seed:            seed,
+		retentionPeriod: do.retentionPeriod,
+		peers:           make(map[string][]peer.AddrInfo),
+		timestamps:      make(map[string]time.Time),
+	}
+}
+
+// Advertise implements discovery.Advertiser and discovery.Discovery.
+func (c *client) Advertise(ctx context.Context, ns string, opts ...discovery.Option) (time.Duration, error) {
+	req := AdvertiseRequest{
+		Namespace: ns,
+	}
+	var res AdvertiseResponse
+
+	_, err := c.rc.Call(ctx, c.seed.ID, MethodAdvertise, req, &res)
+	if err != nil {
+		c.logger.Error("failed to advertise",
+			"err", err,
+		)
+
+		return 0, fmt.Errorf("failed to advertise: %w", err)
+	}
+
+	return res.TTL, nil
+}
+
+// FindPeers implements discovery.Discoverer and discovery.Discovery.
+func (c *client) FindPeers(ctx context.Context, ns string, opts ...discovery.Option) (<-chan peer.AddrInfo, error) {
```

Comment:
Is it ok for `FindPeers` to block until the peers are found? Or would it be better if it returned immediately but peers would be sent over the channel? Didn't check what the caller expects, but if it should be sync it seems weird that they would use a channel for the return value.

Reply:
No, it should be async. Will fix.